### PR TITLE
Read JAVA_OPTIONS env var before OPENJ9_JAVA_OPTIONS

### PR DESCRIPTION
**Pull request by pshipton**
Also fix bug when -XX:+UseNoGC is used in an environment variable.

Fixes https://github.com/eclipse/openj9/issues/8641

### CODE REVIEW

**Comment by pshipton** [comment]
jenkins test sanity,extended,sanity.openjdk zlinux jdk8

**Comment by pshipton** [comment]
jenkins test sanity,extended,sanity.openjdk zlinux jdk11

File: test/functional/Java8andUp/src/org/openj9/test/vmArguments/VmArgumentTests.java [hunk 1]
```diff
@@ -334,7 +335,23 @@ public void testArgEncodingInIbmJavaOptions() {
...
+		Map<String, String> env = pb.environment();
+		String javaOptionsArg = "-Dtest.name=testJavaOptions"; //$NON-NLS-1$
+		env.put(JAVA_OPTIONS, javaOptionsArg);
+		ArrayList<String> actualArguments = runAndGetArgumentList(pb);
+		HashMap<String, Integer> argumentPositions = checkArguments(actualArguments, new String[] {javaOptionsArg});
+		assertTrue(MISSING_ARGUMENT+javaOptionsArg, argumentPositions.containsKey(javaOptionsArg));
+		/* environment variables should come after implicit arguments */
+		assertTrue(JAVA_OPTIONS+ SHOULD_COME_AFTER+DJAVA_HOME, 
+				argumentPositions.get(javaOptionsArg).intValue() > argumentPositions.get(DJAVA_HOME).intValue()); 
+	}
```
Comment (DanHeidinga) [inline test/functional/Java8andUp/src/org/openj9/test/vmArguments/VmArgumentTests.java:351]:
```suggestion
		if (!isIbm)
			ProcessBuilder pb = makeProcessBuilder(new String[] {}, CLASSPATH);
			Map<String, String> env = pb.environment();
			String javaOptionsArg = "-Dtest.name=testJavaOptions"; //$NON-NLS-1$
			env.put(JAVA_OPTIONS, javaOptionsArg);
			ArrayList<String> actualArguments = runAndGetArgumentList(pb);
			HashMap<String, Integer> argumentPositions = checkArguments(actualArguments, new String[] {javaOptionsArg});
			assertTrue(MISSING_ARGUMENT+javaOptionsArg, argumentPositions.containsKey(javaOptionsArg));
			/* environment variables should come after implicit arguments */
			assertTrue(JAVA_OPTIONS+ SHOULD_COME_AFTER+DJAVA_HOME, 
					argumentPositions.get(javaOptionsArg).intValue() > 	argumentPositions.get(DJAVA_HOME).intValue());
		}
	}
```

File: test/functional/Java8andUp/src/org/openj9/test/vmArguments/VmArgumentTests.java [hunk 2]
```diff
@@ -666,30 +683,36 @@ public void testCommandlineArgumentsWithLeadingAndTrailingSpaces() {
 		}
 	}
 
-	/* IBM_JAVA_OPTIONS should take priority over JAVA_TOOL_OPTIONS */
+	/* IBM_JAVA_OPTIONS should take priority over JAVA_TOOL_OPTIONS, and JAVA_OPTIONS */
 	@Test
 	public void testEnvironmentVariableOrdering() {
 		ProcessBuilder pb = makeProcessBuilder(new String[] {}, CLASSPATH);
```
Comment (DanHeidinga) [inline test/functional/Java8andUp/src/org/openj9/test/vmArguments/VmArgumentTests.java:689]:
```suggestion
		if (isIbm) {
			/* disable this test for IBM builds */
			return;
		}
		ProcessBuilder pb = makeProcessBuilder(new String[] {}, CLASSPATH);
```

File: runtime/redirector/redirector.c [hunk 3]
```diff
@@ -463,6 +464,62 @@ parseMemorySizeValue(char *option)
...
 
+#define GC_POLICY_OPTION "-Xgcpolicy:"
+#define LENGTH_GC_POLICY_OPTION (sizeof(GC_POLICY_OPTION) - 1)
+
+static void
+checkEnvOptions(char *envOptions, int *gcPolicy, char **xcompressedstr, char **xnocompressedstr, char **xjvmstr, int *xjvm, char **namedVM, size_t *nameLength, char **xmxstr)
+{
+	char *gcPolicyString = NULL;
+
+	gcPolicyString = findStartOfMostRightOption(envOptions, GC_POLICY_OPTION);
```
Comment (keithc-ca) [inline runtime/redirector/redirector.c:475]: Why not fold this initialization into the declaration?

File: runtime/redirector/redirector.c [hunk 4]
```diff
@@ -491,69 +548,24 @@ chooseJVM(JavaVMInitArgs *args, char *retBuffer, size_t bufferLength)
...
-	const char *gcPolicyOption = "-Xgcpolicy:";
-	size_t gcPolicyOptionLength = strlen(gcPolicyOption);
-	char *gcPolicyString = NULL;
 
 	char *xmxstr = NULL;
 	U_64 requestedHeapSize = 0;
 
-	/* the command line is handled below but look into the ENVVAR_OPENJ9_JAVA_OPTIONS here, since it is a special case */
+	/* 
+	 * The command line is handled below but look into the *JAVA_OPTIONS here, since it is a special case.
```
Comment (keithc-ca) [inline runtime/redirector/redirector.c:556]: Is the `*` in `*JAVA_OPTIONS` significant?

**Comment by pshipton** [comment]
I've addressed the review comments. I tested the changes in https://ci.eclipse.org/openj9/view/Test/job/Grinder/1079/ and via internal build 456583.

File: runtime/redirector/redirector.c (hunk 3 above)
Comment (pshipton) [reply runtime/redirector/redirector.c]: Updated.

File: runtime/redirector/redirector.c (hunk 4 above)
Comment (pshipton) [reply runtime/redirector/redirector.c]: Yes. I've rewritten the comment to be explicit.

**Comment by pshipton** [comment]
I left the original commit as-is, although I rebase it. The changes are in the new commit. I can squash after review.

File: runtime/redirector/redirector.c [hunk 5]
```diff
@@ -463,6 +464,60 @@ parseMemorySizeValue(char *option)
...
+checkEnvOptions(char *envOptions, int *gcPolicy, char **xcompressedstr, char **xnocompressedstr, char **xjvmstr, int *xjvm, char **namedVM, size_t *nameLength, char **xmxstr)
+{
+	char *gcPolicyString = findStartOfMostRightOption(envOptions, GC_POLICY_OPTION);
+	if (NULL == gcPolicyString) {
+		if (hasEnvOption(envOptions, "-XX:+UseNoGC")) {
+			gcPolicyString = "nogc";
+		}
+	}
+	if (NULL != gcPolicyString) {
+		parseGCPolicy(gcPolicyString + LENGTH_GC_POLICY_OPTION, gcPolicy);
```
Comment (DanHeidinga) [inline runtime/redirector/redirector.c:480]: I know this isn't new code but it looks like we index off the end of the buffer here if `gcPolicyString = "nogc"`.

Likely fixable by setting 
```
gcPolicyString = "-Xgcpolicy:nogc"
```

**Comment by DanHeidinga** [comment]
Looks good assuming the pre-existing bug is fixed as well.  Once you squash the commits I'll launch the builds

File: runtime/redirector/redirector.c (hunk 5 above)
Comment (pshipton) [reply runtime/redirector/redirector.c]: Updated.

**Comment by pshipton** [comment]
I'll let Keith take a look before squashing.

**Review by DanHeidinga — APPROVED** [review]
lgtm

File: runtime/redirector/redirector.c [hunk 6]
```diff
@@ -463,6 +464,60 @@ parseMemorySizeValue(char *option)
...
+		xnocompressed = 0;
+		*xnocompressedstr = "-XX:-UseCompressedOops";
+	}
+	
+	*xjvmstr = strstr(envOptions, "-Xjvm:");
+	if (NULL != *xjvmstr) {
+		char *space = NULL;
+
+		xjvm = 0;
+		*namedVM = strstr(envOptions, "-Xjvm:") + 6;
```
Comment (keithc-ca) [inline runtime/redirector/redirector.c:505]: I think this second call to `strstr()` can be replaced by `*xjvmstr`.

File: test/functional/Java8andUp/src/org/openj9/test/vmArguments/VmArgumentTests.java [hunk 7]
```diff
@@ -130,9 +132,9 @@
 	protected static Logger logger = Logger.getLogger(VmArgumentTests.class);
 
 	static {
-		boolean isIbm = System.getProperty("java.vm.vendor").equals("IBM Corporation");
+		isIBM = System.getProperty("java.vm.vendor").equals("IBM Corporation");
 		mandatoryArgumentsJava8 = new String[] {
-				isIbm?XOPTIONSFILE:null,
+				isIBM?XOPTIONSFILE:null,
```
Comment (keithc-ca) [inline test/functional/Java8andUp/src/org/openj9/test/vmArguments/VmArgumentTests.java:137]: nit: spaces missing around `?` and `:` (here and on line 153).

File: test/functional/Java8andUp/src/org/openj9/test/vmArguments/VmArgumentTests.java [hunk 8]
```diff
@@ -666,30 +688,46 @@ public void testCommandlineArgumentsWithLeadingAndTrailingSpaces() {
...
+		env.put(JAVA_TOOL_OPTIONS, javaToolOptionsArg);
 		env.put(OPENJ9_JAVA_OPTIONS, openJ9JavaOptionsArg);
 		ArrayList<String> actualArguments = runAndGetArgumentList(pb);
 		HashMap<String, Integer> argumentPositions = checkArguments(actualArguments, 
-				new String[] {ibmJavaOptionsArg, javaToolOptionsArg, openJ9JavaOptionsArg});
+				new String[] {ibmJavaOptionsArg, javaOptionsArg, javaToolOptionsArg, openJ9JavaOptionsArg});
 		assertTrue(MISSING_ARGUMENT+ibmJavaOptionsArg, 
 				argumentPositions.containsKey(ibmJavaOptionsArg));
+		if (!isJava8 || !isIBM) {
+			assertTrue(MISSING_ARGUMENT+javaOptionsArg,
```
Comment (keithc-ca) [inline test/functional/Java8andUp/src/org/openj9/test/vmArguments/VmArgumentTests.java:710]: nit: spaces missing around `+` (here and several places below).

**Comment by pshipton** [comment]
Made the requested updates.

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
This is ready to be squashed. I'll let @DanHeidinga take it from here.

**Comment by pshipton** [comment]
Squashed, keeping the bug fix separate.

**Comment by DanHeidinga** [comment]
jenkins test sanity,extended,sanity.openjdk zlinux jdk8,jdk11

**Pull request merged**